pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,520 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	cmg "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var (
+	mgr       MetricsManager
+	srv       *http.Server
+	wg        *sync.WaitGroup
+	statusMap map[int]string = map[int]string{
+		0: "Success",
+		1: "Failure",
+		2: "Unknown",
+	}
+)
+
+const (
+	httpPattern = "/metrics"
+	addr        = "localhost:12134"
+	httpAddr    = "http://" + addr + httpPattern
+)
+
+type fakeOpStatus struct {
+	statusCode int
+}
+
+func (s *fakeOpStatus) String() string {
+	if str, ok := statusMap[s.statusCode]; ok {
+		return str
+	}
+	return "Unknown"
+}
+
+func initMgr() {
+	wg = &sync.WaitGroup{}
+	wg.Add(1)
+	mgr = NewMetricsManager()
+	var err error
+	if srv, err = mgr.StartServing(httpPattern, addr, nil, wg); err != nil {
+		log.Fatalf("failed to start serving [%v]", err)
+	}
+}
+
+func shutdown() {
+	if err := srv.Shutdown(context.Background()); err != nil {
+		panic(err)
+	}
+	wg.Wait()
+}
+
+func TestMain(m *testing.M) {
+	initMgr()
+	c := m.Run()
+	shutdown()
+	os.Exit(c)
+}
+
+func TestNew(t *testing.T) {
+	if mgr == nil {
+		t.Errorf("failed testing new")
+	}
+}
+
+func TestStartServing(t *testing.T) {
+	if _, err := http.Get(httpAddr); err != nil {
+		t.Errorf("failed to get response from server %v", err)
+	}
+}
+
```

Comment:
adding one more to cover " recording when no operations have been started", the rest of the cases should have. already been covered in TestConcurrency

Reply:
It would be nice to have simpler test cases for those simple corner cases rather than having to wade through a complicated TestConcurrency failure.

I'm thinking of the perspective from the future when some change elsewhere breaks these tests, and we no longer remember how any of this works :)

Reply:
added simple tests as suggested